### ISSUE TITLE
VGA driver works in an inconsistent way outside of QEMU/Bochs?

### ISSUE THREAD

**Issue by PF94** [byteduck/duckOS#12]
![image](https://user-images.githubusercontent.com/45898787/193675506-faabf509-09f0-4669-bf22-ec617e3e8614.png)

```[94m[kinit] [INFO] Starting duckOS...[39;49m
[90m[kinit] [DEBUG] BIOS boot disk: 0x8[39;49m
[90m[MemoryManager] [DEBUG] Adding memory region from 0x0 -> 0x9efff (Unused, Unreserved)[39;49m
[90m[MemoryManager] [DEBUG] Ignoring too-small memory region at 0x9fc00[39;49m
[90m[MemoryManager] [DEBUG] Adding memory region from 0xf0000 -> 0xfffff (Used, Reserved)[39;49m
[90m[MemoryManager] [DEBUG] Adding memory region from 0x100000 -> 0x1ffeffff (Unused, Unreserved)[39;49m
[90m[MemoryManager] [DEBUG] Adding memory region from 0x1fff0000 -> 0x1fffffff (Used, Unreserved)[39;49m
[90m[MemoryManager] [DEBUG] Adding memory region from 0xfec00000 -> 0xfec00fff (Used, Reserved)[39;49m
[90m[MemoryManager] [DEBUG] Adding memory region from 0xfee00000 -> 0xfee00fff (Used, Reserved)[39;49m
[90m[MemoryManager] [DEBUG] Adding memory region from 0xfffc0000 -> 0xffffffff (Used, Reserved)[39;49m
[90m[Device] [DEBUG] Device 1,5 registered[39;49m
[90m[Device] [DEBUG] Device 1,8 registered[39;49m
[90m[Device] [DEBUG] Device 1,3 registered[39;49m
[90m[I8042] [DEBUG] Attempting to initialize...[39;49m
[90m[I8042] [DEBUG] Found dual-channel controller[39;49m
[90m[I8042] [DEBUG] Keyboard available[39;49m
[90m[I8042] [DEBUG] Mouse available[39;49m
[93m[I8042] [WARN] Read timed out...[39;49m
[91m[I8042] [ERROR] Error enabling keyboard[39;49m
[93m[I8042] [WARN] Read timed out...[39;49m
[91m[I8042] [ERROR] Error enabling mouse[39;49m
[90m[Device] [DEBUG] Device 5,2 registered[39;49m
[90m[Device] [DEBUG] Device 1,16 registered[39;49m
[94m[CommandLine] [INFO] Command line options: ''[39;49m
[90m[Device] [DEBUG] Device 29,0 registered[39;49m
[94m[VGA] [INFO] Found a bochs-compatible VGA device at 0:2.0[39;49m
[90m[VGA] [DEBUG] virtual framebuffer mapped from 0xe0000000 to 0xc0261000[39;49m
[94m[kinit] [INFO] Debug mode is enabled.[39;49m
[90m[kinit] [DEBUG] First stage complete.[39;49m
[90m[TaskManager] [DEBUG] Initializing tasking...[39;49m
[90m[kinit] [DEBUG] Tasking initialized.[39;49m
[90m[Device] [DEBUG] Device 4,0 registered[39;49m
****
```

**Comment by byteduck** [comment]
Weird - it should be using the framebuffer passed to it by GRUB. Looks like it's probably a memory layout issue, judging by the garbage on the screen. It did work on real hardware and virtualbox at one point, but it's been  while since I've tested.

**Comment by PF94** [comment]
It appears to be inconsistent, since sometimes it does boot to the UI but then it later crashes when opening something.

![image](https://user-images.githubusercontent.com/45898787/193680926-6e99fc92-a837-46ac-9ee5-7536d2a2b6ec.png)

**Comment by PF94** [comment]
![image](https://user-images.githubusercontent.com/45898787/193684655-56f1f1c8-c0a9-45be-a285-049b3cbfbe86.png)
On 86box (which emulates old computers from as late as '98/'99), this issue also occurs. In this setup, I'm emulating the "S3 ViRGE GX2/Diamond Sleath 3D 4000". However, I can't debug the attempts due to 86box having no serial support (or something like that).

**Comment by byteduck** [comment]
Yeah... Looks like some sort of repeated data structure is being written to where the video memory should be, which is probably an issue with the memory management.

I know for a fact that the `kstd::map` implementation is incorrect, as it stands. I was working on a fix for that before I had to take a hiatus from this project over the summer due to an internship which required an NDA that prevented me from working on duckOS, so I never got around to fixing that. I'm back in school now, so while I have the freedom to fix this stuff now, I don't have the time :) I've been slowly chipping away at adding a unit testing framework to the kernel to verify this kind of stuff. 

I also want to fix that incorrect hashmap implementation, and probably overhaul the memory system altogether while I'm at it... duckOS has largely been a learning process for me, so large parts of the kernel are very old and very bad.

TL;DR: Probably memory issues. Working on overhauling this stuff.

**Comment by byteduck** [comment]
The memory subsystem has been overhauled, so this _might_ be fixed now. I can't repro it, so I'll mark this as closed for now unless it happens again.
